docs/books/valhalla-mjolnir/src/03_graph_construction_process.md

```diff
@@ -0,0 +1,310 @@
+
+```cpp
+// From valhalla/mjolnir/graphbuilder.h
+class GraphBuilder {
+public:
+  static void Build(const boost::property_tree::ptree& pt,
+                   const std::string& ways_file,
+                   const std::string& way_nodes_file,
+                   const std::string& nodes_file,
+                   const std::string& edges_file,
+                   const std::string& complex_restriction_file,
+                   const std::string& osmdata_file);
+};
+```
+
+The implementation in `src/mjolnir/graphbuilder.cc` performs these key steps:
+
+1. **Node Creation**: Convert OSM nodes to graph nodes
+2. **Edge Creation**: Convert OSM ways to directed edges
+3. **Initial Attribution**: Assign basic attributes to edges (road class, surface type, etc.)
+4. **Connectivity Establishment**: Connect edges at nodes to form a network
+
+Here's a simplified version of the process:
+
+```cpp
+// From src/mjolnir/graphbuilder.cc
+void GraphBuilder::Build(...) {
+  // Create a map of OSM node ids to graph node ids
+  std::unordered_map<uint64_t, uint32_t> node_map;
+  
+  // Create nodes
+  for (const auto& osm_node : osm_nodes) {
+    if (IsIntersection(osm_node)) {
+      uint32_t node_id = AddNode(osm_node);
+      node_map[osm_node.id] = node_id;
+    }
+  }
+  
+  // Create edges
+  for (const auto& osm_way : osm_ways) {
+    // For each segment in the way
+    for (size_t i = 0; i < osm_way.nodes.size() - 1; i++) {
+      uint64_t from_node_id = osm_way.nodes[i];
+      uint64_t to_node_id = osm_way.nodes[i+1];
+      
+      // Create directed edge(s)
+      AddEdge(node_map[from_node_id], node_map[to_node_id], osm_way);
+      
+      // If not one-way, add edge in reverse direction
+      if (!osm_way.one_way) {
+        AddEdge(node_map[to_node_id], node_map[from_node_id], osm_way);
+      }
+    }
+  }
+}
+```
+
+During this process, Mjolnir identifies which OSM nodes should become graph nodes. Typically, these are:
+- Intersections where multiple ways meet
+- Endpoints of ways
+- Points where road attributes change significantly
+
+Not every OSM node becomes a graph node, as this would create an unnecessarily large graph. Instead, intermediate nodes along a way are used to define the shape of edges but don't become actual graph nodes.
+
+## Graph Enhancement
+
+After the initial graph is built, Mjolnir enhances it with additional information using the `GraphEnhancer` class:
+
+```cpp
+// From valhalla/mjolnir/graphenhancer.h
+class GraphEnhancer {
+public:
+  static void Enhance(const boost::property_tree::ptree& pt,
+                     const std::string& access_file);
+};
+```
+
+The enhancement process includes:
+
+1. **Adding Turn Lanes**: Information about lane configuration at intersections
+2. **Speed Assignment**: Setting speed limits based on road class and other factors
+3. **Access Restrictions**: Determining which modes of travel can use each edge
```

Comment:
nit: the description is too broad, we determine 99.9% of accesibility up front at parse time however there are more complicated access restrictions that depend on things like a vehicles weight or width etc ie not mode of transport but other characteristics. thats what this code does it adds these special restrictions to the edges where they belong